packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java

```diff
@@ -5,42 +5,207 @@
 package io.flutter.plugins.sharedpreferences;
 
 import android.content.Context;
+import android.content.SharedPreferences;
+import android.util.Base64;
+import android.util.Log;
+import androidx.annotation.NonNull;
+import androidx.annotation.VisibleForTesting;
 import io.flutter.embedding.engine.plugins.FlutterPlugin;
 import io.flutter.plugin.common.BinaryMessenger;
-import io.flutter.plugin.common.MethodChannel;
+import io.flutter.plugins.sharedpreferences.Messages.SharedPreferencesApi;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
 
 /** SharedPreferencesPlugin */
-public class SharedPreferencesPlugin implements FlutterPlugin {
-  private static final String CHANNEL_NAME = "plugins.flutter.io/shared_preferences_android";
-  private MethodChannel channel;
-  private MethodCallHandlerImpl handler;
+public class SharedPreferencesPlugin implements FlutterPlugin, SharedPreferencesApi {
+  static final String TAG = "SharedPreferencesPlugin";
+
+  SharedPreferencesListEncoder listEncoder;
+
+  private static final String SHARED_PREFERENCES_NAME = "FlutterSharedPreferences";
+
+  // Fun fact: The following is a base64 encoding of the string "This is the prefix for a list."
+  private static final String LIST_IDENTIFIER = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBhIGxpc3Qu";
+  private static final String BIG_INTEGER_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBCaWdJbnRlZ2Vy";
+  private static final String DOUBLE_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBEb3VibGUu";
+
+  private android.content.SharedPreferences preferences;
+
+  public SharedPreferencesPlugin() {
+    this(new ListEncoder());
+  }
+
+  @VisibleForTesting
+  SharedPreferencesPlugin(SharedPreferencesListEncoder testListEncoder) {
+    listEncoder = testListEncoder;
+  }
 
   @SuppressWarnings("deprecation")
-  public static void registerWith(io.flutter.plugin.common.PluginRegistry.Registrar registrar) {
+  public static void registerWith(
+      @NonNull io.flutter.plugin.common.PluginRegistry.Registrar registrar) {
     final SharedPreferencesPlugin plugin = new SharedPreferencesPlugin();
-    plugin.setupChannel(registrar.messenger(), registrar.context());
+    plugin.setUp(registrar.messenger(), registrar.context());
+  }
+
+  private void setUp(@NonNull BinaryMessenger messenger, @NonNull Context context) {
+    preferences = context.getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE);
```

Comment:
Which means this will crash in `onDetachedFromEngine`. You'll want to null out prefs in that case.

Which introduces a threading problem actually, because code using it could be running on the task queue. You'll need to make all the other methods that use `preferences`:
- Make a local variable that uses preferences (I don't remember the shadowing rules for Java; if you can, doing `SharedPreferences preferences = this.preferences` to avoid accidentally using the member variable would be good.)
- Null check that local before proceeding. You can return some dummy value in the null case since if it's detached from the engine it's not going to matter what you return.

Reply:
Resolving since I misread the call.

Thinking about nulling out `preferences` more, hopefully it's fine to keep using the preferences object even if the context we got it from is gone; I don't see anything in the docs saying it's not. In that case, we should be able to leave it as-is, and avoid dealing with thread races.

We'll want to keep an eye out for any reports of exceptions at shutdown (e.g., if something throws saying we're using a `SharedPreferences` instance that's no longer valid or something) that would indicate that's not correct.